### PR TITLE
2 bugfixes and several new features for createclass, createcontent and createrole handlers

### PR DESCRIPTION
**Pull request by bchoquet-heliopsis**
Hi,
here are a few updates I made :
## CreateClass Handler
- Handling manual attribute placement instead of handling it only in new classes
- Attribute names are serialized later in process allowing update
- Class attributes are removed when replacing class
- Class attributes can now be set as information collectors
- Bug fix : when no class name is specified there is no change. (Previously first attribute name was used)
- Bug fix : when no urlAliasPattern was specified, it was replaced with a blank string
- Deleting an unnecessary and (IMHO) ill thought special case for ezboolean datatype default value
## CreateRole Handler
- Mapping class identifier to class ID for Class and ParentClass Limitations
- Mapping section identifier to section ID for Section Limitation
## CreateContent Handler
- Handling ezobjectrelationlist attributes content as a comma separated reference ids list
- Datatype content mapping falls back to fromString method instead of displaying error messages
- Handling node priorities either on an autoincrement basis or with fixed values
- Handling node sort array
- Allowing to set complete xml code instead of a single paragraph
## CreateSection handler
- new XML attribute `sectionIdentifier`

### CODE REVIEW

**Comment by bchoquet-heliopsis** [comment]
Just added one more commit implementing ezsection identifier attribute in 2 places :
- in create section handler identifiers can be set
- in create role handler section identifier may be used for limitation value

**Comment by nfrp** [comment]
Very nice, thanks for that Benjamin.
We'll look into this very soon.
Cheers,

File: classes/ezxmlinstallerhandler.php [hunk 1]
```diff
@@ -147,6 +147,43 @@ class eZXMLInstallerHandler
         return $string;
     }
 
+    /**
+     * Browses node to replace any string references in subnodes or attributes
```
Comment (bdunogier) [inline classes/ezxmlinstallerhandler.php]: "Browses $node..."

File: xmlinstallerhandler/ezcreateclass.php [hunk 2]
```diff
@@ -50,7 +58,7 @@ class eZCreateClass extends eZXMLInstallerHandler
 
             $this->writeMessage( "\tClass '$classIdentifier' will be updated.", 'notice' );
 
-            $classURLAliasPattern   = is_object( $class->getAttribute( 'urlAliasPattern' ) ) ? $class->getAttribute( 'urlAliasPattern' ) : null;
+            $classURLAliasPattern   = $class->getAttribute( 'urlAliasPattern' ) ? $class->getAttribute( 'urlAliasPattern' ) : null;
```
Comment (bdunogier) [inline xmlinstallerhandler/ezcreateclass.php:61]: I guess this means that $class->getAttribute( 'urlAliasPattern' ) is either false or an object, right ?

File: xmlinstallerhandler/ezcreateclass.php [hunk 3]
```diff
@@ -115,9 +125,9 @@ class eZCreateClass extends eZXMLInstallerHandler
                         $class->setAttribute( 'is_container', $classIsContainer );
                         $class->setAttribute( 'url_alias_name', $classURLAliasPattern );
 
-                        $class->set();
+                        //$class->set();
```
Comment (bdunogier) [inline xmlinstallerhandler/ezcreateclass.php]: if it needs to be removed, remove it, don't comment it :-)

File: xmlinstallerhandler/ezcreateclass.php [hunk 4]
```diff
@@ -228,7 +238,7 @@ class eZCreateClass extends eZXMLInstallerHandler
 
                 $params = array();
                 $params['identifier']               = $attributeIdentifier;
-                $params['serialized_name_list']     = $classAttributeNameList->serializeNames();
+                $params['name_list']     = $classAttributeNameList;
```
Comment (bdunogier) [inline xmlinstallerhandler/ezcreateclass.php]: make sure alignment is maintained

File: xmlinstallerhandler/ezcreatecontent.php [hunk 5]
```diff
@@ -458,6 +545,22 @@ class eZCreateContent extends eZXMLInstallerHandler
         }
     return false;
     }
+
+    /**
+     * Updates node priority
```
Comment (bdunogier) [inline xmlinstallerhandler/ezcreatecontent.php]: "Updates the priority for node $node_id to $priority"

This way you don't need to re-document them on the @param line (mentionned in the coding standards)

File: classes/ezxmlinstallerhandler.php (hunk 1 above)
Comment (bchoquet-heliopsis) [inline classes/ezxmlinstallerhandler.php]: fixed in 7075f665

File: xmlinstallerhandler/ezcreateclass.php (hunk 3 above)
Comment (bchoquet-heliopsis) [inline xmlinstallerhandler/ezcreateclass.php]: Fixed in 7075f665

File: xmlinstallerhandler/ezcreateclass.php (hunk 4 above)
Comment (bchoquet-heliopsis) [inline xmlinstallerhandler/ezcreateclass.php]: Fixed in 7075f665

File: xmlinstallerhandler/ezcreatecontent.php (hunk 5 above)
Comment (bchoquet-heliopsis) [inline xmlinstallerhandler/ezcreatecontent.php]: Fixed in 7075f665

File: xmlinstallerhandler/ezcreateclass.php (hunk 2 above)
Comment (bchoquet-heliopsis) [inline xmlinstallerhandler/ezcreateclass.php:61]: A string is not an object :)

**Comment by bchoquet-heliopsis** [comment]
Most of the changes are relative to a single feature and not interdependant, the easiest way to review it should be commit by commit. 
Below is a short explanation regarding each commit or couple of commits, I hope it'll help you.

I'll come up later with a test case for commit 01cb3b33459dc5f906b3 which was a little tricky and, as I can see now, poorly documented.

The general idea behind all these proposed updates is to use ezxmlinstaller to configure an ez instance data model without having to rely on numeric ids which vary between environments.

# CreateClass :

## ad5a36d and 02c891f : handling of attribute placement

prior to these commits, the `placement` xml attribute was simply not used. You can now add a new content object attribute in a specific position or update an existing content object attribute position

## 226a6055 : replace reference id strings in datatype parameters

you could not use references in datatype parameters (for example setting an object relation default node placement) but had to use a numeric ids

## 5b72977d : remove class attributes when replacing class and ebf8898 : implementing infocollection when creating attributes

I guess those two are self explanatory

## 5b72977d and b9b14480     : bug when not specifyinc class names

previous code looked for any descendant `Name` XML element which would replace existing class name by the first declared class attribute's name if no class name was defined. New behavior is to keep existing class name if not redefined.

## 1f63319 : let ezboolean type handle its default value

content class attributes parameters are handled by datatype class's `unserializeContentClassAttribute()` method. `eZBooleanType` already looks for a default value in the `DatatypeParameters` element. This useless chunk of code messed with me for too long so I mercilessly got rid of it.

# CreateRole :

## 4fcaedf1 : handling of `Class` and `ParentClass` limitation type

a way to use human readable values in when specifing class related limitations

# CreateContent :

## 81adaf21 : handling object relation list datatype + datatype default behavior

not all datatypes could be used when creating content. `ezobjectrelationlist` can now be input as comma separated list of references. I also thought the `eZContentObjectAttribute::fromString()` method was the appropriate fallback. 
BTW this switch is a pain in the ass, a strategy pattern would be welcome ;)

## b841b44f and d6d7027c    : Handling node priorities and sort array

new nodes sort array and priorities could no be set

## ee35c3d : Allowing to set full xmltext

original behavior was useful when creating a single paragraph but annoying when creating more complex content. It now can be disabled

# Misc :

## 0bd8174 : updating new fetaures version numbers

comments updating. Feel free to give me a new version

## 41c8d9d : Implementing section identifier attribute

I just discovered this human readable attribute

## 62d23c9 : backward compatibility with translation system

we use ezxmlinstaller to maintain old sites as well so a little BC was needed

**Comment by bchoquet-heliopsis** [comment]
I recalled the reason behind 01cb3b3, no need for a test case :

Original code basically executes the following code when updating a class attribute (see https://github.com/ezsystems/ezxmlinstaller/blob/master/xmlinstallerhandler/ezcreateclass.php#L449) :

``` php
$classAttributeName = serialize( $nameList );
$classAttribute->setAttribute( 'serialized_name_list', $classAttributeName  );
$classAttribute->store();
```

It cannot work because
1. `eZContentClassAttribute` unserializes its names in an internal property `$NameList` on initialization and reserializes on store (https://github.com/ezsystems/ezpublish/blob/master/kernel/classes/ezcontentclassattribute.php#L291)
2. no setter is defined on attribute `serialized_name_list` to update`$NameList`

So the workaround was to keep an instance of `eZSerializedObjectNameList`, set it to the `$NameList` and let `store()` serialize it. The better way would obviously be to have a setter function deserialize new value but I'm not sure it can be implemented in an eZpersistentObject definition on an existing attribute...

File: xmlinstallerhandler/ezcreateclass.php [hunk 6]
```diff
@@ -476,7 +484,21 @@ class eZCreateClass extends eZXMLInstallerHandler
...
     }
 
-
+    /**
+     * Updates placement for each attribute in a class instance
+     *
+     * @since 1.2.1
+     * @param eZContentClass $class class instance to update
+     */
+    protected function _adjustClassAttributesPlacement(eZContentClass $class)
```
Comment (andrerom) [inline xmlinstallerhandler/ezcreateclass.php]: No need to have underscore in function name, it is already protected.

Comment (bchoquet-heliopsis) [inline xmlinstallerhandler/ezcreateclass.php]: Underscores removed.

File: xmlinstallerhandler/ezcreaterole.php [hunk 7]
```diff
@@ -130,6 +130,15 @@ class eZCreateRole extends eZXMLInstallerHandler
         $limitationValue = $this->getReferenceID( $limitationValue );
         switch( $limitationType )
         {
+            case 'Class':
+            case 'ParentClass':
+                $class = eZContentClass::fetchByIdentifier( $limitationValue );
```
Comment (xc) [inline xmlinstallerhandler/ezcreaterole.php]: Shouldn't it have a is_int here? For my understanding there are class_id and class_identifier both supported.

Comment (bchoquet-heliopsis) [inline xmlinstallerhandler/ezcreaterole.php]: You're right, it is now corrected

**Pull request merged**